### ISSUE TITLE
Feature request: Allow passing a hard-coded hostname or IP to match TLS/X509 certificates against

### ISSUE THREAD

**Issue by exikyut** [TigerVNC/tigervnc#1441]
**Is your feature request related to a problem? Please describe.**
TigerVNC shows a non-optional security warning when the certificate returned by a server doesn't match the host or IP being connected to.

https://github.com/TigerVNC/tigervnc/blob/7fe85cd4a7ce89b55662fd445b78d50001bad68a/common/rfb/CSecurityTLS.cxx#L361-L366

This can happen in scenarios where
- the destination might not have a DNS address and has a dynamic (or DHCP-set) IP
- a reverse connection is being made from a machine that obviously is not going to have address in the server's cert

**Describe the solution you'd like**
I'd like to be able to pass a commandline or config-file option to specify a comma-delimited set of free-form strings to be passed to `gnutls_x509_crt_check_hostname()` in sequence (succeeding on the first match), to be considered with equal precedence to the string representation of the host/IP being connected to.

Perhaps the host/IP can be tested/matched first; perhaps also the matching string can be reported in the log (I'm not sure at what debug level).

**Describe alternatives you've considered**
- In one dimension one might consider disabling the popup and relegating the warning to stdout or a log file. Something something "`curl -k` considered harmful" etc. No. :)
- It's possible only passing a single string will prove to be perfectly adequate. I'm trying to envisage obscure scenarios involving load-balancers or multiple worker/node machines in HPC that might return different certificates. Given that I'm the first person reporting this it's perhaps possible I'm completely overthinking the engineering here :)

**Additional context**
It's possible this is the only viable solution to handle the reverse connection edge-case when using X509Vnc (https://github.com/TigerVNC/tigervnc/issues/1440).

**Comment by CendioOssman** [comment]
This is very non-standard so I'd rather not go this route. I prefer if we look at how browsers handle this as they have a lot more experience dealing with certificates. And I've never seen something like this for them.

Browsers don't have a reverse connection mode though, so if we narrow the coverage this might be more interesting. Basically a list of acceptable reverse connection machines.

**Comment by CendioOssman** [comment]
Hmm... SNI is also an issue. We are supposed to provide a single name as part of the TLS handshake. So if we have a list then which one to use?

Perhaps not an issue in practice and we can just not provide any name at that point.

**Comment by aussielunix** [comment]
Hitting this issue too.
Running Ubuntu 22.04 in AWS on EC2 with a private IP.
VNC connections are to a public IP (EIP) that DNAT's to the ec2 instance.


```bash
dpkg -l | grep -i tigerv
ii  tigervnc-common                            1.12.0+dfsg-4                           amd64        Virtual network computing; Common software needed by servers
ii  tigervnc-standalone-server                 1.12.0+dfsg-4                           amd64        Standalone virtual network computing server
ii  tigervnc-tools                             1.12.0+dfsg-4                           amd64        Virtual network computing; Tools for both clients and servers

```

**Comment by CendioOssman** [comment]
Maybe we are overcomplicating this. Don't the browser store an exception for incorrect name as well? If we do the same, then this problem should mostly go away.

**Comment by aussielunix** [comment]
I tried a few things.

* edit openssl.cnf and make the CN the AWS EIP
  * `vncveiwer 54.xxx.xxx.184:5901`
  * **I still get the warning**
* edit openssl.cnf and make the commonName = @HostName@
* add an entry for the ec2 instance's hostname to my local /etc/hosts
  ```bash
  54.xxx.xxx.184 ip-172-xx-1-1xx.ap-southeast-2.compute.internal
  ```
* vncviewer connect to the ec2 instances hostname
  * `vncveiwer ip-172-xx-1-1xx.ap-southeast-2.compute.internal:5901`
* **the warning goes away**

Even when I manage get the CN in the generated certificate correct, I still get the warning as if something serverside is trying to compare it to the fqdn of the instance (which will never match in AWS EC2)

**Comment by CendioOssman** [comment]
The proper modern place for the name of a system is in something called "Subject Alternative Name", not in common name. Could you try putting the external IP address there? If that doesn't work, then there is a bug somewhere.

**Comment by aussielunix** [comment]
I still get a popup telling me the hostname does not match the server certificate.


```bash
# /etc/tigervnc/openssl.cnf

commonName     = @HostName@
...
...
...
subjectAltName = @alt_names

[alt_names]
DNS.1   = 54.xx.xx.184
DNS.2   = @HostName@
DNS.3   = localhost
DNS.4   = 172.xx.x.1xx
```

```bash
# openssl x509 -in .vnc/ip-172-xx-x-1xx.ap-southeast-2.compute.internal-SrvCert.pem -text
...
...
Issuer: CN = ip-172-xx-x-1xx.ap-southeast-2.compute.internal
...
...
Subject: CN = ip-172-xx-x-1xx.ap-southeast-2.compute.internal
...
...
X509v3 Authority Key Identifier: 
  DirName:/CN=ip-172-xx-x-1xx.ap-southeast-2.compute.internal
...
  X509v3 Subject Alternative Name: 
    DNS:54.xx.xx.184, DNS:ip-172-xx-x-1xx.ap-southeast-2.compute.internal, DNS:localhost, DNS:172.xx.x.1xx
```

**Comment by CendioOssman** [comment]
That's not necessary, as I'm fortunate enough to get paid to do this. But you are welcome to add a bounty to some issue you find interesting. We're always interested in ways to attract more developers.

**Comment by CendioOssman** [comment]
It should hopefully be fixed sufficiently in fc3ff6c84c7d578dd13fec783aef7de44964d726. It will now remember an exception for an invalid hostname, so you just need to approve it manually once.
